src/sentry/spans/buffer_v2.py

```diff
@@ -0,0 +1,290 @@
+                    p.zrangebyscore(
+                        key, 0, cutoff, start=0 if max_segments else None, num=max_segments or None
+                    )
+                    p.zcard(key)
+
+                result = iter(p.execute())
+
+        segment_ids = []
+        queue_sizes = []
+
+        with metrics.timer("sentry.spans.buffer.flush_segments.step2"):
+            with self.client.pipeline(transaction=False) as p:
+                # ZRANGEBYSCORE output
+                for segment_span_ids in result:
+                    # process return value of zrevrangebyscore
+                    for segment_id in segment_span_ids:
+                        # TODO: SSCAN
+                        segment_ids.append(segment_id)
+                        p.smembers(segment_id)
+
+                    # ZCARD output
+                    queue_sizes.append(next(result))
+
+                segments = p.execute()
+
+        for shard_i, queue_size in zip(self.assigned_shards, queue_sizes):
+            metrics.timing(
+                "sentry.spans.buffer.flush_segments.queue_size",
+                queue_size,
+                tags={"shard_i": shard_i},
+            )
+
+        return_segments = {}
+
+        for segment_id, segment in zip(segment_ids, segments):
+            return_segment = set()
+            metrics.timing("sentry.spans.buffer.flush_segments.num_spans_per_segment", len(segment))
+            for payload in segment:
+                return_segment.add(payload)
+
+            return_segments[segment_id] = return_segment
+        metrics.timing("sentry.spans.buffer.flush_segments.num_segments", len(return_segments))
+
+        return sum(queue_sizes), return_segments
+
+    def done_flush_segments(self, segment_ids: dict[SegmentId, set[bytes]]):
+        payload_nums = []
+        metrics.timing("sentry.spans.buffer.done_flush_segments.num_segments", len(segment_ids))
+        with metrics.timer("sentry.spans.buffer.done_flush_segments"):
+            with self.client.pipeline(transaction=False) as p:
+                for segment_id, payloads in segment_ids.items():
+                    hrs_key = b"span-buf:hrs:" + segment_id
+                    p.get(hrs_key)
+                    p.delete(hrs_key)
+                    p.delete(segment_id)
+
+                    # parse trace_id out of SegmentId, then remove from queue
+                    segment_id_parts = segment_id.split(b":")
+                    project_id = segment_id_parts[2][1:]
+                    trace_id = segment_id_parts[3][:-1]
+                    redirect_map_key = b"span-buf:sr:{%s:%s}" % (project_id, trace_id)
+                    shard = self.assigned_shards[int(trace_id, 16) % len(self.assigned_shards)]
+                    p.zrem(f"span-buf:q:{shard}".encode("ascii"), segment_id)
+
+                    payload_nums.append(len(payloads))
+                    for payload in payloads:
+                        span_id = rapidjson.loads(payload)["span_id"]
```

Comment:
Would it be possible to avoid this load by getting the span_id from somewhere else?

Reply:
we don't store individual span IDs in redis, only payloads, so when we flush we eventually have to get it out of the payload one way or another.